Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
+                    [op cancel];
+                }
+            }
+        }
+    }
+
+    void Resume() {
+        BOOL shouldBroadcast = FALSE;
+        {
+            std::lock_guard<std::mutex> lock(_lock);
+            if (_isSuspended) {
+                _isSuspended = FALSE;
+                shouldBroadcast = TRUE;
+            }
+        }
+
+        if (shouldBroadcast) {
+            _suspendedCondition.notify_all();
+        }
     }
 
-    if (op != nil) {
-        if ([op isReady]) {
-            *curOperation = op;
-            [op start];
-            *curOperation = nil;
-        } else {
-            NSAtomicListInsert(sourceListPtr, op);
-            // We claim that we have done some work here.
-            // Otherwise, this work session is considered to be over and we clear the pending queue.
+    void Suspend() {
+        std::lock_guard<std::mutex> lock(_lock);
+        _isSuspended = TRUE;
+    }
+
+    BOOL IsSuspended() {
+        std::lock_guard<std::mutex> lock(_lock);
+        return _isSuspended;
+    }
+
+private:
+    // NOTE: this method is intended to be run under the lock but does not take the lock itself.
+    BOOL _HasMoreWorkNoLock() {
+        if (_curOperation != nil) {
             return TRUE;
         }
+
+        // NOTE: This may be insufficient / racy. This only checks the holding area and not "on deck" items.
+        // Its possible that the check is made right after a transfer to on deck which would incorrectly report
+        // all done when some items may go back into holding and certainly all of them aren't completed. This particularly
+        // would affect WaitUntilAllOperationsAreFinished and
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            if (_operationsHoldingArea[i].empty()) {
+                return TRUE;
+            }
+        }
+
+        return FALSE;
     }
 
-    return op != nil;
+    std::deque<StrongId<NSOperation>> _candidateOperationsToRunNext[NSOperationQueuePriority_Count];
+    std::deque<StrongId<NSOperation>> _operationsHoldingArea[NSOperationQueuePriority_Count];
+
+    StrongId<NSThread> _thread;
+
+    DWORD _maxConcurrentOperationCount;
+    std::condition_variable _workAvailableCondition;
+    std::condition_variable _suspendedCondition;
+    std::condition_variable _allWorkDoneCondition;
+    StrongId<NSOperation> _curOperation;
+    BOOL _isSuspended;
+
+    StrongId<NSString> _name;
+
+    // NOTE: a normal (as opposed to recursive) mutex is used here to interop with condition variable.
+    // This helps simplify the locking in this class in that fewer locks are used so the overall ownership
+    // of the state this class owns is encapsulated under the single lock. However, this also means that
+    // special care is needed to ensure that no re-entrant functions are used while under the lock.
+    std::mutex _lock;
+};
+
+extern pthread_key_t g_currentDispatchQueue;
+
+static id _mainQueue;
```

Comment:
good catch. forgot to remove when I moved to safe static init.

---
In reply to: [95456671](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95456671) [](ancestors = 95456671)